clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/ChainResolver.java

```diff
@@ -131,4 +133,65 @@ public ChainResolver(final OperationsCodec<K, V> codec, Expiry<? super K, ? supe
       return new ResolvedChain.Impl<>(chain, key, result, 0, expirationTime);
     }
   }
+
+  @SuppressWarnings("unchecked")
+  public Chain compact(Chain chain, long now) {
+    Map<K, PutOperation<K, V>> compacted = new HashMap<>(1);
+    for (Element element : chain) {
+      ByteBuffer payload = element.getPayload();
+      Operation<K, V> operation = codec.decode(payload);
+      K key = operation.getKey();
+      final Result<V> previousResult = compacted.get(key);
+      final Result<V> result = operation.apply(previousResult);
+      if(result == null) {
+        compacted.remove(key);
+      } else if (expiry == Expirations.noExpiration()) {
+        if (result instanceof PutOperation<?, ?>) {
+          compacted.put(key, (PutOperation<K, V>) result); //unchecked
+        } else {
+          compacted.put(key, new PutOperation<>(key, result.getValue(), -Long.MAX_VALUE));
+        }
+      } else if (operation.isExpiryAvailable()) {
+          long expirationTime = operation.expirationTime();
+          if (now >= expirationTime) {
+            compacted.remove(key);
+          } else if (result instanceof PutOperation<?, ?>) {
```

Comment:
There are not much cut&paste anymore apart from a tiny one here. My favor would be to put the different if contents little methods will nice names